[Pokergame step5] 포커게임 결과 화면

corykim0829 opened the pull request:
- 스택뷰를 활용하여 player와 게임타입에 따라서 스택뷰에 카드를 추가해주는 방식을 사용했습니다.
  - 카드들을 담는 스택뷰
  - 카드스택뷰와 이름을 담는 스택뷰
  - 전체 Player, Dealer를 담는 스택뷰
- Dealer와 Player를 대등한 관계에서 처리해주기 위해서 protocol을 만들어서 채택하게 했습니다.
- Card 클래스는 참조가 필요없다고 판단하여 struct로 변경하였습니다.
- View의 구현은 다른 클래스로 만들어 구현하였습니다.

corykim0829 committed:
Refactor: Remove print

corykim0829 committed:
Add: Card image files

corykim0829 committed:
Feat: enum class PlayerCount

corykim0829 committed:
Refactor: Class PokerGame initializer

corykim0829 committed:
Refactor: CardDeck 프로퍼티 접근제어자 변경

corykim0829 committed:
Feat: Method returning numberOfHands in current game type

corykim0829 committed:
Feat: struct Card property - imageFileName

corykim0829 committed:
Feat: UISegmentedControl for GameType

- Add Target to change gameType
- Default gameType is 7-card stud

corykim0829 committed:
Refactor: gameTypeSegementedControl name

corykim0829 committed:
Refactor: enum class PlayerCount case

corykim0829 committed:
Feat: playerCount UISegmentedControl

- Add Target to it

corykim0829 committed:
Feat: StackView contains segmentedControls

corykim0829 committed:
Feat: pokerGameStackView

- Create 3 StackViews to display players cards

corykim0829 committed:
Feat: playerLabel textColor, pokerGame StackView constraint

corykim0829 committed:
Fix: Retrieving Card image filename

corykim0829 committed:
Feat: Reset the game whenever game setting value get changed

- refactoring setupPokerGame method name
- fix playerCount switch in segment action method
- remove unnecessary constraints code

corykim0829 committed:
Refactor: reset methods & constraints

- pokerGameStackView spacing
- cardStackView spacing
- pokerGameStackView constrainsts

corykim0829 committed:
Feat: pokerGameStackView constraints

corykim0829 committed:
Refactor: resetPlayers adopting Playable protocol

corykim0829 committed:
Refactor: stackView default value

corykim0829 committed:
Refactor: CardImageView

corykim0829 committed:
Refactor: PlayerStackView, CardStackView to Views group

corykim0829 committed:
Refactor: PokerGameStackView

corykim0829 committed:
Refactor: pokerGame property value

corykim0829 committed:
Refactor: SegmentedControls, CardStackView

corykim0829 committed:
Refactor: segnentedControlStackView name

corykim0829 committed:
Docs: Update README

corykim0829 committed:
Refactor: CardImageView class commonInit()

corykim0829 committed:
Refactor: Avoid hard-coding in ViewController

corykim0829 committed:
Feat: Shake motion to reset

corykim0829 committed:
Refactor: Typo of class name

corykim0829 committed:
Refactor: TitledSegmentedControl class name

godrm commented on CardGameApp/CardGameApp/Models/Card.swift:
```diff
@@ -57,6 +70,7 @@ extension Card: CustomStringConvertible {
         default:
             rank = String(self.rank.rawValue)
         }
```
여기에 있는 suit, rank에 대한 switch-case 구문은 
Suit나 Rank enum 타입이 CustomStringConvertible을 채택하는 구조가 좋습니다. 
각 타입별로 할 일을 타입 내부에 구현하고, 상위 객체 Card에서는 그걸 조합해서 활용하게 되는거죠.

corykim0829 replied:
여기서 suit와 rank에 CustomStringConvertible을 채택해서 구현한다면, suit, rank 안에서 사용되는 switch-case 구문은 피할 수 없는 부분인게 맞나요?

godrm replied:
네. enum 에서는 어쩔수 없이 매칭이 필요해서 하나는 생깁니다.

godrm commented on CardGameApp/CardGameApp/Models/Dealer.swift:
```diff
@@ -8,7 +8,9 @@
 
-class Dealer {
+class Dealer: Playable {
+    var hands: [Card] = []
```
hands가 여전히 public 이네요. 늘 우선적으로 private var로 선언하세요.

corykim0829 replied:
네 수정하였습니다

godrm commented on CardGameApp/CardGameApp/Models/Dealer.swift:
```diff
@@ -8,7 +8,9 @@
+    
     private let gameType: GameType
     private var deck = CardDeck()
     private var _communityCards: [Card] = []
```
딜러가 갖고 있는 communityCards 가 hands 랑 같은 의미 아니었던가요?
그리고 왜 _ communityCards 만 `_`를 붙였나요?

corykim0829 replied:
getter로 만들었던 건데, communityCards를 없애고 hand로 통일했습니다

godrm commented on CardGameApp/CardGameApp/Models/PokerGame.swift:
```diff
@@ -28,15 +28,20 @@ class PokerGame {
+    
+    func numberOfHands() -> Int {
+        return self.rawValue
+    }
```
enum 타입 뿐만 아니라 내부 속성을 이렇게 getter 형식으로 리턴하는 경우보다는
내부에서 값을 사용해서 로직을 처리하는 메소드를 구현하는 게 좋습니다. 
이 값을 가져가서 어떻게 활용하는 지 고려해보세요.

```
enum GameType: Int {
case sevenCardsStud = 7, fiveCardsStud = 5

func forEach(_ transform : () -> ()) {
    for _ in 0..<self.rawValue {
        transform()
    }
}

}
```

이렇게 선언하면 호출할 때는 `gameType.numberOfHands { }` 형태가 됩니다.
참고하세요.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -9,46 +9,101 @@
-        stackView.translatesAutoresizingMaskIntoConstraints = false
+    private var gameType: GameType = .sevenCardsStud
+    private var playerCount: PlayerCount = .two
+    private var pokerGame: PokerGame = PokerGame(game: .sevenCardsStud, numberOfPlayers: .two)
```
PokerGame() 인스턴스를 만들 때도 gameType, playerCounter 같은 변수를 활용하려고 변수를 선언한거 아닌가요?
위에 변수 선언과 값을 그냥 넘기는 게 일치하지 않을 수도 있나요?

corykim0829 replied:
lazy var 사용을 피해서 private 변수로 만드려고 했는데, 생각해보니 ViewController여서 private로 만들필요가 없기도 한 것 같네요.

godrm replied:
ViewController도 private으로 만드는 건 기본입니다 :)

corykim0829 replied:
그럼 여기서는 lazy var를 사용하는게 더 적절하나요?

godrm replied:
lazy var가 필요하다면 선택할 수도 있는거죠. 왜 필요한건가요?

corykim0829 replied:
말씀하신대로 이미 만들어진 변수 gameType, playerCounter를 활용해서 PokerGame 인스턴스 인자에 넣기 위해서 lazy var를 사용하려고 합니다.

godrm replied:
꼭 리터럴 방식으로 초기화를 해야 하는 건 아닙니다. 
그래서 init() 함수가 있는거니까요. 거기서 초기화를 해도 되죠.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -9,46 +9,101 @@
+            let cardStackView = CardStackView(of: player)
+            
+            playerStackView.addArrangedSubview(cardStackView)
+            pokerGameStackView.addArrangedSubview(playerStackView)
```
pokerGame 내부 속성 players를 가져와서 여기서 추가하고 탐색하고 있습니다.
PokerGame 내부에서 이 동작들을 처리하고, 
ViewController는 PokerGame 상위 객체니까 pokerGame에게 메소드로 명령하는 게 좋습니다. 
위에 enum 타입에 대한 힌트처럼 PokerGame도 Players 배열을 하나씩 탐색해서 처리하는 게 아니라, ViewController가 처리할 동작을 클로저 형식으로 넘기도록 작성해보세요.

godrm commented on CardGameApp/CardGameApp/ViewController.swift:
```diff
@@ -9,46 +9,101 @@
+            playerCount = .four
+        default:
+            break
+        }
```
handleGameTypeSegmentChanged 나 handlePlayerCountSegmentChanged 에서 사용하는 
switch-case 구문도 각 enum 타입의 생성자 메소드를 활용할 수 있는게 좋습니다. 
guard, if, switch 비교문을 줄일 수 있는 방법을 고민해보세요.

godrm commented on CardGameApp/CardGameApp/Views/CardImageView.swift:
```diff
@@ -0,0 +1,32 @@
+    required init?(coder: NSCoder) {
+        super.init(coder: coder)
+        commonInit()
+    }
```
커스텀 View 객체를 초기화할 때는 중요한 포인트가 몇 개 있습니다. 
그 중에서도 init(frame:) 과 init(coder:) 두 메소드가 매우 중요합니다. 
어느 생성 초기화 메소드로 만들어도 동일하게 동작하도록 구현하는 게 좋습니다. 
commonInit()을 시도한 부분은 좋습니다. 다만 init(frame:) 생성자가 빠져있네요.

corykim0829 replied:
아하 그럼 만약 파라미터 사용이 필요한 생성자도 사용하려면 총 생성자가 3개가 만들어지게 되는건가요??

godrm commented on CardGameApp/CardGameApp/Views/PokerGameStackViews.swift:
```diff
@@ -0,0 +1,84 @@
+                addArrangedSubview(CardImageView(named: filename))
+            })
+    }
+}
```
보통은 클래스 이름이 다르면 다른 파일로 분리하는 게 좋습니다. 
다른 클래스가 변경됐는데, 여기 파일 모든 클래스를 다시 컴파일해서 다른 파일에 영향을 주게 됩니다.

godrm commented on CardGameApp/CardGameAppTests/PokerGameTests.swift:
```diff
@@ -16,7 +16,7 @@ class PokerGameTests: XCTestCase {
-        let pokerGame = PokerGame(game: .sevenCardsStud, numberOfPlayers: 4)
+        let pokerGame = PokerGame(game: .sevenCardsStud, numberOfPlayers: .four)
         
         XCTAssert(pokerGame.players.count == 4)
```
테스트할 때 내부 속성을 확인해서 테스트하기 보다는 메소드를 실행하고, 리턴 결과를 확인하는 형태로 테스트를 작성하는 게 좋습니다.
내부 속성을 확인하는 형태로 만들다보면, 상위 모듈에서 값을 가져와서 처리하는 형태의 코드가 만들어지게 됩니다. 지양하세요.

godrm reviewed (commented):
각 세부 설정을 하는 뷰 클래스를 분리한 것은 좋은 시도네요.
다음 피드백을 확인해보세요.

corykim0829 committed:
Refactor: enum GameType, PlayerCount initializer

- Avoiding swtich-case

corykim0829 committed:
Refactor: Access modifier pokerGame in VC

corykim0829 committed:
Refactor: enum rank, suit switch-case

corykim0829 committed:
Refactor: default value of PokerGame initializer parameters

corykim0829 committed:
Refactor: Class files

- Each class has a own swift file
- Remove break lines

corykim0829 committed:
Refactor: PokerGame, Player, Dealer

- Dealer inherits Player
- Every class or struct doesn't have public properties
- Pass Subobject's properties using closure parameter

corykim0829 committed:
Refactor: PokerGame, Player, ViewController

- Not accessing properties of subobjects in ViewController
- Add name property in Player

corykim0829 committed:
Refactor: Remove comments

godrm commented on CardGameApp/CardGameApp/Models/Card.swift:
```diff
@@ -44,33 +57,6 @@ extension Card: Equatable {
-        }
-        
-        return "\(suitString)\(rank)"
+        return "\(String(describing: self.suit))\(String(describing: self.rank))"
```
CustomStringConvertible 을 구현한 타입은 문자열로 그대로 바꿀 수 있습니다. 
`return "\(suit)\(rank)"` 형태만 해도 될 겁니다. 참고하세요.

godrm commented on CardGameApp/CardGameApp/Models/Player.swift:
```diff
@@ -8,14 +8,25 @@
+    }
+    
+    func forEachName(_ handler: (String) -> ()) {
+        handler(name)
```
아니 이름은 배열도 아닌데 굳이 클로저로 만들 필요는 없죠 ㅎㅎ 
그냥 private(set) var name 정도만 해도 됩니다. 모든 것을 다 클로저로 만들어야 하는 건 아닙니다.

corykim0829 replied:
저는 ViewController에서 직접 이름을 사용해야해서 이렇게 사용했는데, 이렇게 안하면 하위객체 접근이 힘들지 않나요?

godrm replied:
private 으로 하라는 것은 외부에서 값을 바꾸지 못하게 하려는 게 첫번째 이유구요
배열이 아니라 값인 경우, 바꾸지 못하고 read-only로 읽을 수만 있는 것도 괜찮습니다. 
연습하는 과정이니까 private 으로 하라고 한건데 private(set)가 차선책이 됩니다.

godrm commented on CardGameApp/CardGameApp/Models/PokerGame.swift:
```diff
@@ -10,38 +10,74 @@ import Foundation
+    
+    func forEachPlayer(_ handler: (Int, Player) -> ()) {
+        players.enumerated().forEach { (i, player) in
+            handler(i, player)
```
이 코드는 for player in players 나 큰 차이가 없네요. 클로저 연습하는 게 아니라면 꼭 이렇게 만들어야 하는 건 아닙니다. 
어차피 위에 setupPlayer에서 부르는 거니까요

corykim0829 replied:
원래 ViewController에서 사용하려고 만든건데, 만들어져 있는거라 PokerGame에서도 사용했습니다. 하위 객체를 사용하지 않을 때 사용하면 혼동이 올 수도 있을 것 같네요!

godrm replied:
맞아요. public 이면 더 혼란스럽니다. 메소드만 보고 어떤 동작을 위한 것인지 알기 어려우니까요
클로저를 쓴다고 코드가 꼭 좋아지는 게 아니라는 것을 명심하세요.
